haigha/channel.py

```diff
@@ -227,10 +235,25 @@ def process_frames(self):
         Process the input buffer.
         '''
         while len(self._frame_buffer):
+            # It would make sense to call next_frame, but it's
+            # technically faster to repeat the code here.
+            frame = self._frame_buffer.popleft()
+
+            if self._emergency_close_pending:
+                # Implement stability rule from AMQP 0.9.1 section 1.5.2.5.
+                # Method channel.close: "After sending this method, any
+                # received methods except Close and Close-OK MUST be discarded."
+                #
+                # NOTE: presently, we limit our implementation of the rule to
+                # the "emergency close" scenario to avoid potential adverse
+                # side-effect during normal user-initiated close
+                if (not isinstance(frame, MethodFrame) or
+                      frame.class_id != 20 or
+                      frame.method_id not in (40, 41)):
+                    self.logger.warn("Emergency channel close: dropping input "
```

Comment:
I don't like that this breaks the abstraction layer, wherein the class id and method id is defined between the `Connection` and the individual `ProtocolClass` es. I can see where that could be challenging to resolve.